polar-core/src/kb.rs

```diff
@@ -90,196 +102,338 @@ impl KnowledgeBase {
+        return prototype_fields
             .fields
             .iter()
-            .map(|(k, template_value)| {
+            .map(|(k, prototype_value)| {
                 rule_fields
                     .fields
                     .get(k)
-                    .map(|rule_value| rule_value == template_value)
+                    .map(|rule_value| rule_value == prototype_value)
                     .unwrap_or_else(|| false)
             })
             .all(|v| v);
     }
 
-    fn rule_params_match(&self, rule: &Rule, template: &Rule) -> bool {
-        for (rule_param, template_param) in rule.params.iter().zip(template.params.iter()) {
-            let res = match (
-                template_param.parameter.value(),
-                template_param.specializer.as_ref().map(Term::value),
+    /// Check that a rule parameter that has a pattern specializer matches a prototype parameter that has a pattern specializer.
+    fn check_pattern_param(
+        &self,
+        index: usize,
+        rule_pattern: &Pattern,
+        prototype_pattern: &Pattern,
+    ) -> PolarResult<RuleParamMatch> {
+        Ok(match (prototype_pattern, rule_pattern) {
+            (Pattern::Instance(prototype_instance), Pattern::Instance(rule_instance)) => {
+                // if tags match, all prototype fields must match those in rule fields, otherwise false
+                if prototype_instance.tag == rule_instance.tag {
+                    if self.param_fields_match(
+                        &prototype_instance.fields,
+                        &rule_instance.fields,
+                    ) {
+                        RuleParamMatch::True
+                    } else {
+                        RuleParamMatch::False(format!("Rule specializer {:#?} on parameter {} did not match prototype specializer {:#?} because the specializer fields did not match.", rule_instance, index, prototype_instance))
+                    }
+                // If tags don't match, then rule specializer must be a subclass of prototype specializer
+                } else if let Some(Value::ExternalInstance(ExternalInstance {
+                    instance_id,
+                    ..
+                })) = self
+                    .constants
+                    .get(&prototype_instance.tag)
+                    .map(|t| t.value())
+                {
+                    if let Some(rule_mro) = self.mro.get(&rule_instance.tag) {
+                        if rule_mro.contains(instance_id)
+                            && self.param_fields_match(
+                                &prototype_instance.fields,
+                                &rule_instance.fields,
+                            )
+                        {
+                            RuleParamMatch::True
+                        } else {
+                            RuleParamMatch::False(format!("Rule specializer {} on parameter {} must be a subclass of prototype specializer {}", rule_instance.tag,index, prototype_instance.tag))
```

Comment:
What if it's a subclass, but the parameters don't match?

Reply:
good catch